### PR TITLE
Validate name in .dockstore.yml

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
This PR validates the workflow name in the .dockstore.yml using the same validation that's in the front-end. If the name is invalid, the GitHub release fails and you can view the error message in the GitHub App logs:

Screenshot 2021-10-28 at 08-54-24 Dockstore My Workflows

I also refactored `checkEntryName` into a helper class for string input validation so future string input validation methods that are usable in several places can be put here as well. 

There are a handful of existing GitHub app workflows that have names containing spaces which is invalid with the validation used in this PR. This is a [published workflow](https://dockstore.org/workflows/github.com/arvados/bh20-seq-resource/Pangenome%20Generator:master?tab=info) on prod with a workflow name containing spaces ([corresponding .dockstore.yml](https://github.com/pubseq/bh20-seq-resource/blob/aa1b7e5f507148149f0d65f9bfe86fe9daffbcc8/.dockstore.yml#L3)). With these workflows, any subsequent pushes to these repos would result in a failure. If they fix the workflow name, it would create a new workflow since the workflow path is technically different. 

**Issue**
[SEAB-1866](https://ucsc-cgl.atlassian.net/browse/SEAB-1866)

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] Check the Snyk dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

**Comment by kathy-t** [comment]
Made the error message more specific (indicates what type of entry it is).
Screenshot 2021-10-29 at 11-49-59 Dockstore My Workflows

Created documentation ticket: https://ucsc-cgl.atlassian.net/browse/SEAB-3571

**Review by coverbeck — APPROVED** [review]
I just commented on the JIRA ticket you just created, but I'm wary of being unable to update existing workflows. But if nobody else objects...

**Comment by kathy-t** [comment]
> I just commented on the JIRA ticket you just created, but I'm wary of being unable to update existing workflows. But if nobody else objects...

Charles' comment on the JIRA ticket:

> Proposal: What if we instead don’t enforce the name validation on existing workflows? If any of these workflows in question or in use, we’ve just blocked a way for authors to update the workflows.

Thinking about the behaviour for name validation during manual workflow/tool registration, we don't try go back and fix the invalid ones (I'm not sure if there are any), so the equivalent behaviour here would be to only validate names for new workflows and ignore existing ones. 

I could go either way. It would be a more pleasant user experience if they don't suddenly run into a validation error that wasn't there before, but then their workflow would be an exception to our name validation...but I guess you could argue that  if they never update their GitHub app workflow then it'll always be an exception

**Comment by denis-yuen** [comment]
> I could go either way. It would be a more pleasant user experience if they don't suddenly run into a validation error that wasn't there before, but then their workflow would be an exception to our name validation...but I guess you could argue that if they never update their GitHub app workflow then it'll always be an exception 

I think the only reason we're really doing validation is to avoid future issues. Since these particular existing workflows seem ok, I don't have a strong objection to applying this only to future workflows with the caveat that there may be the occasional edge case where these workflows might look/behave oddly

**Comment by kathy-t** [comment]
Changed it so the name validation only applies to new workflows. Re-requesting reviews, will remove https://github.com/dockstore/dockstore-deploy/wiki/1.12-Release-Checklist#release-notes once it's approved since this is no longer a breaking change

**Review by denis-yuen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java [hunk 1]
```diff
@@ -562,6 +563,8 @@ private Workflow createOrGetWorkflow(Class workflowType, String repository, User
                 throw new CustomWebApplicationException("User does not have an account on Dockstore.", LAMBDA_FAILURE);
             }
 
+            StringInputValidationHelper.checkEntryName(workflowType, workflowName);
```
Comment (david4096) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:566]: If I'm reading this correctly the entry name will be checked only during create. One of my questions coming to the PR was what would happen to existing workflows with names that don't pass validation (exceed length or contain spaces/special characters). I presume there would be no effect?

**Review by david4096 — APPROVED** [review]
Looks good. Sort of wondering if we can reuse this code or nicely merge with what @NatalieEO is doing for username validation. Adding more methods to the same class?

Also raises for me the issue of response sanitization. As noted above, we may stop the addition of "bad" names using this but there are still entries in Dockstore that don't satisfy the conditions.

**Review by coverbeck — APPROVED** [review]

**Review by garyluu — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java (hunk 1 above)
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:566]: Yeah, that's correct. This validation has no effects on existing workflows with invalid names

**Review by NatalieEO — APPROVED** [review]

**Comment by kathy-t** [comment]
> Looks good. Sort of wondering if we can reuse this code or nicely merge with what @NatalieEO is doing for username validation. Adding more methods to the same class?

Yeah, the username validation method could be added to this class if the method can be used in multiple places

> Also raises for me the issue of response sanitization. As noted above, we may stop the addition of "bad" names using this but there are still entries in Dockstore that don't satisfy the conditions.

It seems like the consensus from the discussions in this PR is that it's okay for these existing entries to not satisfy the name validation

**Pull request merged**